src/component/1d/matrix/Stocsy.tsx

```diff
@@ -0,0 +1,128 @@
+import { xFindClosestIndex } from 'ml-spectra-processing';
+import { matrixToStocsy } from 'nmr-processing';
+
+import { useChartData } from '../../context/ChartContext';
+import { useScaleChecked } from '../../context/ScaleContext';
+import { usePanelPreferences } from '../../hooks/usePanelPreferences';
+import { PathBuilder } from '../../utility/PathBuilder';
+import { getYScaleWithRation } from '../utilities/scale';
+
+import { interpolatedColorsPoints, useMatrix } from './useMatrix';
+
+interface StocsyProps {
+  x: Float64Array | never[];
+  y: number[];
+  color: string[];
+  scaleRatio: number;
+}
+
+interface StocsyIndexPointProps {
+  x: Float64Array | never[];
+  color: string[];
+  chemicalShift: number | null;
+  xIndex: number;
+}
+
+export function Stocsy() {
+  const {
+    view: {
+      spectra: { activeTab },
+    },
+  } = useChartData();
+  const options = usePanelPreferences('matrixGeneration', activeTab);
+
+  if (!options) return;
+
+  const { scaleRatio, showStocsy, chemicalShift } = options;
+  if (!showStocsy) return null;
+
+  return <InnerStocsy scaleRatio={scaleRatio} chemicalShift={chemicalShift} />;
+}
+export function InnerStocsy({ scaleRatio, chemicalShift }) {
+  const matrix = useMatrix();
+
+  if (!matrix) return null;
+
+  const { x, matrixY } = matrix;
+  const xIndex = xFindClosestIndex(x, chemicalShift ?? x[0], { sorted: false });
+  const { color, y } = matrixToStocsy(matrixY, xIndex);
```

Comment:
This method, matrixToStocsy, takes 30% of the time and is recalculated currently continuously. This should not be the case if matrixY and xIndex does not change.